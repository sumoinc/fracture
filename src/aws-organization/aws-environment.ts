import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../cdk";

export interface IAwsEnvironment {
  readonly orgId: string;
  readonly account: string;
  readonly region: ValueOf<typeof AwsRegionIdentifier>;
}

export interface AwsEnvironmentOptions {
  readonly orgId: string;
  readonly account: string;
  readonly region: ValueOf<typeof AwsRegionIdentifier>;
}

export class AwsEnvironment extends Component {
  public static all(project: CdkProject): Array<AwsEnvironment> {
    const isDefined = (c: Component): c is AwsEnvironment =>
      c instanceof AwsEnvironment;
    return project.components.filter(isDefined);
  }

  /**
   * Make sure this environment is configured for the organization.
   */
  public static ensureExists(
    project: CdkProject,
    options: AwsEnvironmentOptions
  ): AwsEnvironment {
    /**
     * Find function
     */
    const isDefined = (c: Component): c is AwsEnvironment =>
      c instanceof AwsEnvironment &&
      c.orgId === options.orgId &&
      c.account === options.account &&
      c.region === options.region;

    /**
     * It should exist in the project supplied.
     * Return this instance
     */
    return (
      project.components.find(isDefined) ?? new AwsEnvironment(project, options)
    );
  }

  public readonly orgId: string;
  public readonly account: string;
  public readonly region: ValueOf<typeof AwsRegionIdentifier>;

  constructor(
    public readonly project: CdkProject,
    options: AwsEnvironmentOptions
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId = options.orgId;
    this.account = options.account;
    this.region = options.region;
  }

  public config(): IAwsEnvironment {
    return {
      orgId: this.orgId,
      account: this.account,
      region: this.region,
    };
  }
}
