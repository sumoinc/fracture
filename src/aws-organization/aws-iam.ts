/**
 * AWS Managed policies.
 */
export const IamManagedPolicy = {
  ADMINISTRATOR_ACCESS: "AdministratorAccess",
  BILLING: "Billing",
  DATABASE_ADMINISTRATOR: "DatabaseAdministrator",
  DATA_SCIENTIST: "DataScientist",
  NETWORK_ADMINISTRATOR: "NetworkAdministrator",
  POWER_USER_ACCESS: "PowerUserAccess",
  READON_ONLY_ACCESS: "ReadOnlyAccess",
  SECURITY_AUDIT: "SecurityAudit",
  SUPPORT_USER: "SupportUser",
  SYSTEMK_ADMINISTRATOR: "SystemAdministrator",
  VIEW_ONLY_ACCESS: "ViewOnlyAccess",
} as const;
