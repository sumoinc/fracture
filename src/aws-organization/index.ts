export * from "./aws-account";
export * from "./aws-environment";
export * from "./aws-organization";
export * from "./aws-organization";
export * from "./aws-profile";
export * from "./aws-region";
