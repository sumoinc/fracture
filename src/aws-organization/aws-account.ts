import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsEnvironment, IAwsEnvironment } from "./aws-environment";
import { AwsOrganization } from "./aws-organization";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../cdk";
import { DeployTarget } from "../cdk/config/deploy-target";
import {
  GithubOidcRole,
  GithubOidcRoleOptions,
  IGithubOidcRole,
} from "../github/github-oidc-role";

export interface IAwsAccount {
  readonly orgId: string;
  readonly account: string;
  readonly environments: Array<IAwsEnvironment>;
  //readonly deployTargets: Array<IDeployTarget>;
  readonly githubOidcRoles: Array<IGithubOidcRole>;
}

export interface AwsAccountOptions {
  readonly orgId?: string;
  readonly account: string;
  readonly regions?: Array<ValueOf<typeof AwsRegionIdentifier>>;
  /**
   * Define the of releases allowed into this account.
   */
  readonly githubOidcRoleOptions?: Array<
    Omit<GithubOidcRoleOptions, "account">
  >;
}

export class AwsAccount extends Component implements IAwsAccount {
  public static all(project: CdkProject): Array<AwsAccount> {
    const isDefined = (c: Component): c is AwsAccount =>
      c instanceof AwsAccount;
    return project.components.filter(isDefined);
  }

  /**
   * Create account if needed or return it if not.
   */
  public static ensureExists(
    project: CdkProject,
    options: AwsAccountOptions
  ): AwsAccount {
    /**
     * Find function
     */
    const isDefined = (c: Component): c is AwsAccount =>
      c instanceof AwsAccount && c.account === options.account;

    /**
     * It should exist in the project supplied.
     * Return this instance
     */
    return (
      project.components.find(isDefined) ?? new AwsAccount(project, options)
    );
  }

  public readonly orgId: string;
  public readonly account: string;

  constructor(public readonly project: CdkProject, options: AwsAccountOptions) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId =
      options.orgId ??
      AwsOrganization.byAccount(project, options.account).orgId;
    this.account = options.account;

    options.regions?.forEach((region) => {
      AwsEnvironment.ensureExists(this.project, {
        orgId: this.orgId,
        account: this.account,
        region,
      });
    });
  }

  public get environments() {
    return AwsEnvironment.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public get deployTargets() {
    return DeployTarget.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public get githubOidcRoles() {
    return GithubOidcRole.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public config(): IAwsAccount {
    return {
      account: this.account,
      orgId: this.orgId,
      environments: this.environments.map((e) => e.config()),
      //deployTargets: this.deployTargets.map((t) => t.config()),
      githubOidcRoles: this.githubOidcRoles.map((r) => r.config()),
    };
  }
}
