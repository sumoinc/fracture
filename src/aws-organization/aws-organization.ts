import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsAccount } from "./aws-account";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../projects";

export interface AwsOrganizationOptions {
  readonly orgId: string;
  readonly ssoStartUrl: string;
  readonly ssoAccount: string;

  /**
   * @default AwsRegionIdentifier.US_EAST_1
   */
  readonly ssoRegion?: ValueOf<typeof AwsRegionIdentifier>;
}

export class AwsOrganization extends Component {
  public static all(project: CdkProject): Array<AwsOrganization> {
    const isDefined = (c: Component): c is AwsOrganization =>
      c instanceof AwsOrganization;
    return project.components.filter(isDefined);
  }

  public readonly orgId: string;
  public readonly ssoStartUrl: string;
  public readonly ssoAccount: string;
  public readonly ssoRegion: ValueOf<typeof AwsRegionIdentifier>;

  constructor(
    public readonly project: CdkProject,
    options: AwsOrganizationOptions
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId = options.orgId;
    this.ssoStartUrl = options.ssoStartUrl;
    this.ssoAccount = options.ssoAccount;
    this.ssoRegion = options.ssoRegion ?? AwsRegionIdentifier.US_EAST_1;
  }

  public get accounts() {
    return AwsAccount.all(this.project).filter((a) => a.orgId === this.orgId);
  }

  public hasAccount(account: string): boolean {
    return this.accounts.some((a) => a.account === account);
  }

}
