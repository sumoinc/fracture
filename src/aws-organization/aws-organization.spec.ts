import { AwsOrganization } from "./aws-organization";
import { AwsRegionIdentifier } from "./aws-region";
import { CommonProject } from "../projects/common-project";
import { synthFiles } from "../util";

describe("Success conditions", () => {
  test("Smoke Test", () => {
    const project = new CommonProject({
      name: "my-project",
    });

    // add org
    const org = new AwsOrganization(project, {
      orgId: "o-abc-123",
      ssoStartUrl: "https://example.com",
      ssoAccount: "123456789012",
    });
    expect(org).toBeTruthy();
  });
});

describe("Files", () => {
  test("All generated files", () => {
    const project = new CommonProject({
      name: "my-project",
    });
    const content = synthFiles(project);
    expect(content).toBeTruthy();
    // expect(content).toMatchSnapshot();
    //console.log(JSON.stringify(content, null, 2));
  });

  test("Bootstrapping Environments", () => {
    // base project
    const project = new CommonProject({
      name: "my-project",
    });

    // add org
    const org = new AwsOrganization(project, {
      orgId: "o-abc-123",
      ssoStartUrl: "https://example.com",
      ssoAccount: "123456789012",
    });

    // bootstraps for org
    org.bootstrap({
      account: "000000000000",
      region: AwsRegionIdentifier.US_EAST_1,
    });

    const content = synthFiles(project);
    expect(content).toBeTruthy();
    expect(content[".projen/tasks.json"]).toMatchSnapshot();
    // console.log(JSON.stringify(content[".projen/tasks.json"], null, 2));
  });

  test("With Deploy Profiles", () => {
    // base project
    const project = new CommonProject({
      name: "my-project",
    });

    // add org
    const org = new AwsOrganization(project, {
      orgId: "o-abc-123",
      ssoStartUrl: "https://example.com",
      ssoAccount: "123456789012",
    });

    // bootstraps for org
    org.addDeploymentProfile({
      account: "000000000000",
      region: AwsRegionIdentifier.US_EAST_1,
    });

    const content = synthFiles(project);
    expect(content).toBeTruthy();
    expect(content[".projen/tasks.json"]).toMatchSnapshot();

    // console.log(JSON.stringify(content[".projen/tasks.json"], null, 2));
  });

  test("With Login Profile", () => {
    // base project
    const project = new CommonProject({
      name: "my-project",
    });

    // add org
    const org = new AwsOrganization(project, {
      orgId: "o-abc-123",
      ssoStartUrl: "https://example.com",
      ssoAccount: "123456789012",
    });

    // bootstraps for org
    org.addSsoLogin({
      account: "000000000000",
      region: AwsRegionIdentifier.US_EAST_1,
    });

    const content = synthFiles(project);
    expect(content).toBeTruthy();
    expect(content[".projen/tasks.json"]).toMatchSnapshot();

    //console.log(JSON.stringify(content[".projen/tasks.json"], null, 2));
  });
});
