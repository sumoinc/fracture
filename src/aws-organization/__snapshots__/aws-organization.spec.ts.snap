// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files Bootstrapping Environments 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "bootstrap": Object {
      "description": "Bootstrap all environments.",
      "name": "bootstrap",
      "steps": Array [
        Object {
          "spawn": "bootstrap:000000000000:us-east-1:hnb659fds",
        },
      ],
    },
    "bootstrap:000000000000": Object {
      "description": "Bootstrap environments for 000000000000.",
      "name": "bootstrap:000000000000",
      "steps": Array [
        Object {
          "spawn": "bootstrap:000000000000:us-east-1:hnb659fds",
        },
      ],
    },
    "bootstrap:000000000000:us-east-1": Object {
      "description": "Bootstrap environments for 000000000000:us-east-1.",
      "name": "bootstrap:000000000000:us-east-1",
      "steps": Array [
        Object {
          "spawn": "bootstrap:000000000000:us-east-1:hnb659fds",
        },
      ],
    },
    "bootstrap:000000000000:us-east-1:hnb659fds": Object {
      "condition": "[ ! -n \\"$CI\\" ]",
      "description": "Deploy for 000000000000:us-east-1 with qualifier \\"hnb659fds\\"",
      "name": "bootstrap:000000000000:us-east-1:hnb659fds",
      "steps": Array [
        Object {
          "exec": "npx cdk@latest bootstrap aws://000000000000/us-east-1 --no-bootstrap-customer-key=true --profile=bootstrap-000000000000-us-east-1 --qualifier=hnb659fds",
        },
      ],
    },
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "default",
        },
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bump": Object {
      "condition": "! git log --oneline -1 | grep -q \\"chore(release):\\"",
      "description": "Bumps version based on latest git tag and generates a changelog entry",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "bump",
      "steps": Array [
        Object {
          "builtin": "release/bump-version",
        },
      ],
    },
    "clobber": Object {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": Object {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": Array [
        Object {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in \\"scratch\\" branch",
        },
        Object {
          "exec": "git checkout $BRANCH",
        },
        Object {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        Object {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        Object {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        Object {
          "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
      "steps": Array [
        Object {
          "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
        },
      ],
    },
    "eject": Object {
      "description": "Remove projen from the project",
      "env": Object {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": Array [
        Object {
          "spawn": "default",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern $@ src test build-tools projenrc .projenrc.ts",
          "receiveArgs": true,
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "profiles": Object {
      "description": "Write all profiles.",
      "name": "profiles",
      "steps": Array [
        Object {
          "spawn": "profiles:bootstrap-000000000000-us-east-1",
        },
      ],
    },
    "profiles:bootstrap-000000000000-us-east-1": Object {
      "condition": "[ ! -n \\"$CI\\" ]",
      "description": "Write profile for bootstrap-000000000000-us-east-1",
      "name": "profiles:bootstrap-000000000000-us-east-1",
      "steps": Array [
        Object {
          "exec": "aws configure set sso_start_url https://example.com --profile bootstrap-000000000000-us-east-1 && aws configure set sso_region us-east-1 --profile bootstrap-000000000000-us-east-1 && aws configure set sso_account_id 000000000000 --profile bootstrap-000000000000-us-east-1 && aws configure set sso_role_name AdministratorAccess --profile bootstrap-000000000000-us-east-1 && aws configure set region us-east-1 --profile bootstrap-000000000000-us-east-1",
        },
      ],
    },
    "release": Object {
      "description": "Prepare a release from \\"main\\" branch",
      "env": Object {
        "RELEASE": "true",
      },
      "name": "release",
      "steps": Array [
        Object {
          "exec": "rm -fr dist",
        },
        Object {
          "spawn": "bump",
        },
        Object {
          "spawn": "build",
        },
        Object {
          "spawn": "unbump",
        },
        Object {
          "exec": "git diff --ignore-space-at-eol --exit-code",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "spawn": "eslint",
        },
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "unbump": Object {
      "description": "Restores version to 0.0.0",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "unbump",
      "steps": Array [
        Object {
          "builtin": "release/reset-version",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpx npm-check-updates@16 --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,constructs,eslint-config-prettier,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,prettier,projen,standard-version,ts-jest,ts-node,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser constructs eslint-config-prettier eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit prettier projen standard-version ts-jest ts-node typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;

exports[`Files With Deploy Profiles 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "default",
        },
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bump": Object {
      "condition": "! git log --oneline -1 | grep -q \\"chore(release):\\"",
      "description": "Bumps version based on latest git tag and generates a changelog entry",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "bump",
      "steps": Array [
        Object {
          "builtin": "release/bump-version",
        },
      ],
    },
    "clobber": Object {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": Object {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": Array [
        Object {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in \\"scratch\\" branch",
        },
        Object {
          "exec": "git checkout $BRANCH",
        },
        Object {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        Object {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        Object {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        Object {
          "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
      "steps": Array [
        Object {
          "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
        },
      ],
    },
    "eject": Object {
      "description": "Remove projen from the project",
      "env": Object {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": Array [
        Object {
          "spawn": "default",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern $@ src test build-tools projenrc .projenrc.ts",
          "receiveArgs": true,
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "profiles": Object {
      "description": "Write all profiles.",
      "name": "profiles",
      "steps": Array [
        Object {
          "spawn": "profiles:deployment-000000000000-us-east-1",
        },
      ],
    },
    "profiles:deployment-000000000000-us-east-1": Object {
      "condition": "[ ! -n \\"$CI\\" ]",
      "description": "Write profile for deployment-000000000000-us-east-1",
      "name": "profiles:deployment-000000000000-us-east-1",
      "steps": Array [
        Object {
          "exec": "aws configure set sso_start_url https://example.com --profile deployment-000000000000-us-east-1 && aws configure set sso_region us-east-1 --profile deployment-000000000000-us-east-1 && aws configure set sso_account_id 000000000000 --profile deployment-000000000000-us-east-1 && aws configure set sso_role_name AdministratorAccess --profile deployment-000000000000-us-east-1 && aws configure set region us-east-1 --profile deployment-000000000000-us-east-1",
        },
      ],
    },
    "release": Object {
      "description": "Prepare a release from \\"main\\" branch",
      "env": Object {
        "RELEASE": "true",
      },
      "name": "release",
      "steps": Array [
        Object {
          "exec": "rm -fr dist",
        },
        Object {
          "spawn": "bump",
        },
        Object {
          "spawn": "build",
        },
        Object {
          "spawn": "unbump",
        },
        Object {
          "exec": "git diff --ignore-space-at-eol --exit-code",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "spawn": "eslint",
        },
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "unbump": Object {
      "description": "Restores version to 0.0.0",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "unbump",
      "steps": Array [
        Object {
          "builtin": "release/reset-version",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpx npm-check-updates@16 --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,constructs,eslint-config-prettier,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,prettier,projen,standard-version,ts-jest,ts-node,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser constructs eslint-config-prettier eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit prettier projen standard-version ts-jest ts-node typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;

exports[`Files With Login Profile 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "default",
        },
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bump": Object {
      "condition": "! git log --oneline -1 | grep -q \\"chore(release):\\"",
      "description": "Bumps version based on latest git tag and generates a changelog entry",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "bump",
      "steps": Array [
        Object {
          "builtin": "release/bump-version",
        },
      ],
    },
    "clobber": Object {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": Object {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": Array [
        Object {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in \\"scratch\\" branch",
        },
        Object {
          "exec": "git checkout $BRANCH",
        },
        Object {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        Object {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        Object {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        Object {
          "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
      "steps": Array [
        Object {
          "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
        },
      ],
    },
    "eject": Object {
      "description": "Remove projen from the project",
      "env": Object {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": Array [
        Object {
          "spawn": "default",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern $@ src test build-tools projenrc .projenrc.ts",
          "receiveArgs": true,
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "release": Object {
      "description": "Prepare a release from \\"main\\" branch",
      "env": Object {
        "RELEASE": "true",
      },
      "name": "release",
      "steps": Array [
        Object {
          "exec": "rm -fr dist",
        },
        Object {
          "spawn": "bump",
        },
        Object {
          "spawn": "build",
        },
        Object {
          "spawn": "unbump",
        },
        Object {
          "exec": "git diff --ignore-space-at-eol --exit-code",
        },
      ],
    },
    "sso:login": Object {
      "description": "log into all profiles.",
      "name": "sso:login",
      "steps": Array [
        Object {
          "spawn": "sso:login:sso-login-000000000000-us-east-1",
        },
      ],
    },
    "sso:login:sso-login-000000000000-us-east-1": Object {
      "condition": "[ ! -n \\"$CI\\" ]",
      "description": "SSO Login for sso-login-000000000000-us-east-1",
      "name": "sso:login:sso-login-000000000000-us-east-1",
      "steps": Array [
        Object {
          "exec": "aws sso login --profile \\"sso-login-000000000000-us-east-1\\"",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "spawn": "eslint",
        },
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "unbump": Object {
      "description": "Restores version to 0.0.0",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "unbump",
      "steps": Array [
        Object {
          "builtin": "release/reset-version",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpx npm-check-updates@16 --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,constructs,eslint-config-prettier,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,prettier,projen,standard-version,ts-jest,ts-node,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser constructs eslint-config-prettier eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit prettier projen standard-version ts-jest ts-node typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;
